### ISSUE TITLE
bug: silent wrong results in two score-analytics aggregations (histogram binning & pivot-table averages)

### ISSUE THREAD

**Issue by hassaanch23** [langfuse/langfuse#15208]
## Summary

While differential-testing Langfuse's client-side statistical code against reference libraries (`numpy`, `scikit-learn`), I found two **silent** correctness bugs — they return wrong numbers to users with no error, on the dashboards people use to read eval results:

1. **Numeric score histogram** (`createHistogramData`) counts scores in the **wrong bucket**, because it rounds each value to 2 decimals *before* binning. The rendered distribution disagrees with a standard histogram, and buckets are labeled with ranges that don't contain the values counted in them.
2. **Pivot-table subtotals/grand totals** (`calculateSubtotals`) compute averages as an **average of per-group averages** instead of a count-weighted average, so subtotal/total rows overstate or understate metrics like avg cost/latency whenever groups differ in size.

Both live in correctness-critical analytics paths. I've verified the neighbouring inter-rater stats (`calculateCohensKappa`, `calculateWeightedF1Score`, `calculateOverallAgreement`) against scikit-learn and they are **correct** (0 mismatches over 1,200 random cases) — these two are the outliers.

---

## Bug 1 — Numeric score histogram bins on rounded values

**File:** `web/src/features/dashboard/lib/score-analytics-utils.ts` → `createHistogramData` (rendered by the dashboard `NumericScoreHistogram`, produced in `dashboard-router.ts`).

The value is rounded to 2 decimals before its bin index is computed, while the bin edges are not:

```ts
const min = round(Math.min(...numericScoreValues));            // rounded
const binSize = range / bins || 1;
// ...
const shiftedValue = round(value) - min;                       // value rounded to 2dp
const binIndex = Math.min(Math.floor(shiftedValue / binSize), bins - 1);
```

### Reproduction (vs `numpy.histogram`)

A faithful port of the function was run against `numpy.histogram` (same bin count) over 1,200 random score datasets:

```
uniform 0..1 (n=50)          | value-in-wrong-bucket:185/200  numpy-mismatch:117/200
uniform 0..1 (n=200)         | value-in-wrong-bucket:200/200  numpy-mismatch:199/200
narrow 0.80..0.85 (n=40)     | value-in-wrong-bucket:200/200  numpy-mismatch:200/200
1..5 likert (n=60)           | value-in-wrong-bucket:  0/200  numpy-mismatch:  0/200
...
runs differing from numpy.histogram: 916 / 1200
```

Integer (Likert) scores are unaffected — which is why this has gone unnoticed — but any fractional scores (0–1 confidence, LLM-judge scores, cost, latency) are misbinned.

**Minimal deterministic case** (7 equal-width bins over `[0, 1]`, input `[0, 0.857, 0.86, 1]`):

| value  | bucket it's counted in (current) | bucket that contains it |
| ------ | -------------------------------- | ----------------------- |
| 0.857  | **`[0.86, 1]`** ❌               | `[0.71, 0.86]`          |

`round(0.857) = 0.86`, so `floor(0.86 / (1/7)) = 6` → the **top** bucket, even though `0.857 < 0.86`.

A second symptom of the same rounding: for narrow ranges the 2-decimal **labels collapse**, e.g. `createHistogramData([0.855, 0.857, 0.858, 0.861])` produces buckets labeled `["[0.85, 0.85]", "[0.85, 0.86]"]` — a zero-width bucket and duplicated edges.

### Fix

Bin on the raw value (`Math.floor((value - min) / binSize)`), derive edges from the raw min/max, and round only the **display labels** (with enough precision to keep edges distinct). After the fix the port matches `numpy.histogram` exactly (**0 / 1200** mismatches).

---

## Bug 2 — Pivot-table subtotals average the averages

**File:** `web/src/features/widgets/utils/pivot-table-utils.ts` → `applyAggregation` / `calculateSubtotals` (subtotal & grand-total rows in dashboard pivot-table widgets).

```ts
case "avg":
  return values.reduce((sum, val) => sum + val, 0) / values.length;   // mean of per-row means
```

Each pivot data row is already an aggregate (e.g. `avg_cost` over `count` items). Averaging those per-row averages **unweighted** only equals the true overall average when every group has the same size.

### Reproduction

The repo's own test data (and its docstring example) demonstrates it. `calculateGrandTotals` over:

```
{ count: 10, avg_cost: 0.5 }
{ count: 15, avg_cost: 0.6 }
{ count: 20, avg_cost: 0.2 }
{ count: 25, avg_cost: 0.4 }
{ count:  5, avg_cost: 0.8 }
```

| quantity                              | value      |
| ------------------------------------- | ---------- |
| **shown** (avg of averages) `Σavg/5`  | **0.5**    |
| **correct** (weighted) `Σ(avg·count)/Σcount = 32/75` | **0.4267** |

The dashboard silently overstates average cost by ~**17%** here. The existing test even encodes the wrong value:

```ts
avg_cost: 0.5, // (0.5 + 0.6 + 0.2 + 0.4 + 0.8) / 5
```

(`percentile` subtotals have the same problem and are already flagged as "a reasonable approximation" in the code.)

### Fix

Weight each row's value by its `count` when a count metric is present:
`Σ(avg_i · count_i) / Σ(count_i)`. This makes subtotal/total averages exact for the common case (a count column is present) and falls back to the previous unweighted behavior when no weights are available. The truly exact fix for percentiles is query-level (the database can recompute over raw rows), but count-weighting is a strict improvement.

---

## Methodology

Each function was ported 1:1 to JS and executed against reference implementations:
`numpy.histogram` (Bug 1), and `sklearn.metrics.{cohen_kappa_score, f1_score, accuracy_score}` for the neighbouring categorical stats (verified correct — not bugs). Repro scripts and a proposed patch with unit tests are available; happy to open a PR.

**Comment by Taranum01** [comment]
Opening a PR for this: #15211.

**Histogram fix** (`createHistogramData`): bin on the raw value (no rounding before `floor`), derive edges from the raw min/max, and round only the display labels with 3-decimal precision so narrow-range edges stay visually distinct.

**Pivot-table fix** (`calculateSubtotals`/`calculateGrandTotals`): when a `count_<base>` column is present (the same SQL emits both columns), weight each row's `avg_<base>` by its count — `Σ(avg_i × count_i) / Σ(count_i)`. Falls back to the unweighted mean when no count column is available, so no caller breaks.

Added 4 regression tests in a new `score-analytics-utils.clienttest.ts` and extended `pivot-table-utils.servertest.ts` with 3 new cases. All verified to fail on the un-fixed source (via `git stash`) and pass with the fix. `tsc`, `eslint`, and `prettier` all clean on the touched files.

**Comment by tanmay-devhub** [comment]
Hi, I reviewed the two aggregation paths and the existing proposed fixes. I’d be interested in helping with any remaining edge cases, particularly selecting the correct weight when multiple count metrics exist and handling labels for identical-value histograms.

If the current fixes are not moving forward or a focused follow-up would be useful, I’d be happy to work on it.

**Comment by hassaanch23** [comment]
Flagging some duplicated effort here so a maintainer can cut it short rather than review the same fix twice.

There are two open PRs for this issue, touching the same four files: #15209 (mine, opened alongside this report) and #15211 (@Taranum01's). We landed on the same two fixes independently — bin on raw values and round only the display labels; count-weight the pivot averages. That's convergent rather than anyone duplicating anyone, but only one should be reviewed.

Having read both, they are not strictly better/worse — each is stronger in a different place, and @tanmay-devhub's question above lands squarely on the difference:

**Weight selection (#15211 is better).** My `findWeightMetric` resolves *one* weight metric for the whole metrics list and applies it to every average, so with `["avg_latency", "count_latency", "avg_cost", "count_cost"]` it weights `avg_cost` by `count_latency`. That's fine when both counts are the same row cardinality and wrong when they aren't. #15211 pairs `avg_<base>` with its own `count_<base>`, which is the correct rule and exactly the "multiple count metrics" case @tanmay-devhub raised.

**Label precision (#15209 is better).** #15211 hardcodes 3 decimals; that still collapses adjacent edges once the bin size drops below ~0.001 (narrow score ranges), which reintroduces the degenerate `[0.85, 0.85]` labels in a smaller window. Mine derives precision from the bin size, so it scales with the data.

**Percentile subtotals.** Mine also weights `percentile` rows, not just `avg`. Still an approximation — you can't recover a true percentile from per-group percentiles — but weighted beats unweighted. #15211 leaves percentiles unweighted. Reasonable either way; worth an explicit call rather than an accident.

So the best version is #15211's per-base weight pairing with #15209's bin-size-derived label precision, and a deliberate decision on percentiles.

I'm happy for that to happen in whichever PR you prefer, including closing mine — @Taranum01 clearly did the work independently and shouldn't lose it to an ordering accident. If it's useful I'll push the label-precision helper and the percentile handling onto #15211 as a suggestion or a follow-up commit, whichever @Taranum01 would rather. If you'd rather consolidate into #15209 I'll pull the per-base pairing across with credit.

@tanmay-devhub — the multiple-count-metrics case you named is the real remaining gap, and it's the one thing neither PR fully settles (#15211 handles the common shape; neither defines behavior when a base has an `avg` but no matching `count`). Worth a follow-up once one of these lands.

Just say which PR you want to carry it and I'll do the legwork.

**Comment by samvallad33** [comment]
Wrong results in score-analytics histogram binning and pivot averages that surface as success hide bad aggregates behind green queries. Dashboards then drive decisions on silently corrupted stats. Do you want hard assertions on bin edges and pivot denominators with tests that fail closed?

**Comment by hassaanch23** [comment]
@samvallad33 That's already what #15209 does, so no separate pass is needed:

- **Bin edges:** values are binned on the raw score and the edges come from the raw min/max, so nothing is rounded before `floor`. Rounding happens only in the display labels, at a precision derived from the bin size so narrow ranges don't collapse adjacent edges.
- **Pivot denominators:** subtotals and grand totals weight each row's `avg_<base>` by its count — `Σ(avg_i × count_i) / Σ(count_i)` — instead of averaging the averages.

Both come with regression tests that fail on the unfixed code and pass with the fix.

The open question on this issue isn't test coverage, it's that there are two convergent PRs (#15209 and #15211) and neither has had a maintainer review yet — I compared the two above.

**Comment by samvallad33** [comment]
Hi @hassaanch23! I'm really glad #15209 already bins on raw scores and weights the pivot averages by count instead of just averaging averages. Keeping rounding only for the display labels is definitely the right split. I just hope a maintainer picks between #15209 and #15211 soon so all this great convergent work doesn't get stranded!

By the way, when you're off the clock, I'm building a privacy-first app called Vestige to help beautifully organize life's milestones and personal memories. If you ever need a private digital space for your own history, I'd be honored if you checked it out.
